ARC-V DSP extensions

artemiy-volkov opened the pull request:
See commit descriptions.
Do not merge until the GCC part is published and reviewed.

artemiy-volkov committed:
riscv: remove the arcv.bitrev instruction

The arcv.bitrev instruction actually belongs to its own extension
(XARCVbitrev), rather than XARCVudsp.  Remove it.

Signed-off-by: Artemiy Volkov <artemiy@synopsys.com>

artemiy-volkov committed:
fixup! bfd, gas, binutils: add support for the Xarcvudsp extension

Remove line numbers from x-arcv-udsp-fail.l.

Signed-off-by: Artemiy Volkov <artemiy@synopsys.com>

dobbelaj-snps commented on opcodes/riscv-opc.c:
```diff
@@ -3152,7 +3152,142 @@ const struct riscv_opcode riscv_opcodes[] =
 {"arcv.xvwscjmul.vv",              0, INSN_CLASS_XARCVUDSP, "d,s,t,Xav", MATCH_ARCV_UDSP_XVWSCJMUL_VV, MASK_ARCV_UDSP_XVWSCJMUL_VV, match_opcode, 0},
 {"arcv.xvwscrdot.vv",              0, INSN_CLASS_XARCVUDSP, "d,s,t,Xav", MATCH_ARCV_UDSP_XVWSCRDOT_VV, MASK_ARCV_UDSP_XVWSCRDOT_VV, match_opcode, 0},
 {"arcv.xvwscjrdot.vv",              0, INSN_CLASS_XARCVUDSP, "d,s,t,Xav", MATCH_ARCV_UDSP_XVWSCJRDOT_VV, MASK_ARCV_UDSP_XVWSCJRDOT_VV, match_opcode, 0},
-{"arcv.bitrev",              0, INSN_CLASS_XARCVUDSP, "d,s,t", MATCH_ARCV_UDSP_BITREV, MASK_ARCV_UDSP_BITREV, match_opcode, 0},
```
What happened to the 'arc.bitrev' for  udsp related defines ? ( MATCH_ARCV_UDSP_BITREV )
Should it still exist or is it only part of the XARVBITREV ?

artemiy-volkov replied:
this is now part of the xarcvbitrev extension, on which xarcvudsp depends

dobbelaj-snps replied:
In that case the MATCH/MASK_ARCV_UDSP_BITREV defines should be omitted as well

artemiy-volkov replied:
Fixed, many thanks.

dobbelaj-snps commented on gas/testsuite/gas/riscv/x-arcv-udsp.s:
```diff
@@ -74,9 +74,5 @@
 	arcv.xvwscjrdot.vv t3,t6,t0,e16,m1
 	arcv.xvwscjrdot.vv t6,t0,t3,e16,m1
 
-	arcv.bitrev t0,t3,t6
-	arcv.bitrev t3,t6,t0
-	arcv.bitrev t6,t0,t3
-
```
bitrev must still be part of udsp; better to keep checking for that.

artemiy-volkov replied:
These bitrev tests have been moved to x-arcv-bitrev* files.

dobbelaj-snps reviewed (commented):
Had a rough look at this. Some small comments.

One question: has part of these changes been generated in some way ?

artemiy-volkov commented:
> Had a rough look at this. Some small comments.
> 
> One question: has part of these changes been generated in some way ?

Yes, all of them were.

dobbelaj-snps commented on bfd/elfxx-riscv.c:
```diff
@@ -1292,6 +1292,7 @@ static struct riscv_implicit_subset riscv_implicit_subsets[] =
   {"svbare", "+zicsr", check_implicit_always},
 
   {"xsfvcp", "zve32x",  check_implicit_always},
+  {"xarcvudsp", "+xarcvbitrev", check_implicit_always},
```
Shouldn't we have a test that checks that this relationship is valid ? (aka, ability to use a bitrev instruction in xarcvudsp)

artemiy-volkov replied:
Yes, that test existed before and has now been reinstated.

MichielDerhaeg reviewed (changes requested):
I ran this patch on the lit test we have on the LLVM side and it seems like LLVM doesn't agree on the order of the operands. Which one is incorrect?
[incorrect.txt](https://github.com/user-attachments/files/20625692/incorrect.txt)

I also have some concerns about the script used to generate this, but we don't have to address that in this PR, when you're done on the GCC side.

Alexehv77 reviewed (approved):
looks fine - i see the changes for all instructions are similar and all are supported by tests.

artemiy-volkov committed:
bfd, gas, binutils: add support for the ARC-V DSP extensions

This commit implements the following extensions:

XARCVbitrev v1.5.1
XARCVbitstream v1.5.1
XARCVvdsp v1.5.1
XARCVvcplx v1.5.1
XARCVvsad v1.5.1
XARCVmxmb v1.5.1
XARCVmxmc v1.5.1
XARCVmxmd v1.5.1

Signed-off-by: Artemiy Volkov <artemiy@synopsys.com>

artemiy-volkov commented:
> I ran this patch on the lit test we have on the LLVM side and it seems like LLVM doesn't agree on the order of the operands. Which one is incorrect? [incorrect.txt](https://github.com/user-attachments/files/20625692/incorrect.txt)
> 
> I also have some concerns about the script used to generate this, but we don't have to address that in this PR, when you're done on the GCC side.

LLVM is correct. Fixed.
